pkg/kv/kvserver/kvserverpb/liveness.go

```diff
@@ -41,10 +43,66 @@ func (l *Liveness) IsDead(now time.Time, threshold time.Duration) bool {
 	return !now.Before(deadAsOf)
 }
 
-func (l Liveness) String() string {
+func (l *Liveness) String() string {
 	var extra string
-	if l.Draining || l.Decommissioning {
-		extra = fmt.Sprintf(" drain:%t decom:%t", l.Draining, l.Decommissioning)
+	if l.Draining || l.Membership.Decommissioning() || l.Membership.Decommissioned() {
+		extra = fmt.Sprintf(" drain:%t membership:%s", l.Draining, l.Membership.String())
 	}
 	return fmt.Sprintf("liveness(nid:%d epo:%d exp:%s%s)", l.NodeID, l.Epoch, l.Expiration, extra)
 }
+
+// Decommissioning is a shorthand to check if the membership status is DECOMMISSIONING.
+func (c MembershipStatus) Decommissioning() bool { return c == MembershipStatus_DECOMMISSIONING }
+
+// Decommissioned is a shorthand to check if the membership status is DECOMMISSIONED.
+func (c MembershipStatus) Decommissioned() bool { return c == MembershipStatus_DECOMMISSIONED }
+
+// Active is a shorthand to check if the membership status is ACTIVE.
+func (c MembershipStatus) Active() bool { return c == MembershipStatus_ACTIVE }
+
+func (c MembershipStatus) String() string {
+	switch c {
+	case MembershipStatus_ACTIVE:
+		return "active"
+	case MembershipStatus_DECOMMISSIONING:
+		return "decommissioning"
+	case MembershipStatus_DECOMMISSIONED:
+		return "decommissioned"
+	default:
+		err := "unknown membership status, expected one of [active,decommissioning,decommissioned]"
+		panic(err)
+	}
+}
+
+// ValidateLivenessTransition validates transitions of the liveness record,
+// returning an error if the proposed transition is invalid. Ignoring no-ops,
+// the valid state transitions for Membership are as follows:
+// 	  Decommissioning  => Active
+// 	  Active           => Decommissioning
+// 	  Decommissioning  => Decommissioned
+//
+// See diagram above the Membership type for more details.
+func ValidateLivenessTransition(old, new Liveness) error {
+	if old.Membership == new.Membership {
+		// No-op.
+		return nil
+	}
+
+	if new.Membership.Active() && !old.Membership.Decommissioning() {
+		err := fmt.Sprintf("can only recommission a decommissioning node; n%d found to be %s",
+			new.NodeID, old.Membership.String())
+		return status.Error(codes.FailedPrecondition, err)
+	}
+	if new.Membership.Decommissioning() && !old.Membership.Active() {
```

Comment:
But decommissioning is still idempotent? Unclear to me when this is called then

Reply:
This is actually dead code, I had comment to that effect earlier but removed it. We no-op away operations at the caller when observed state is the same as the target state.

```go
	} else if oldLivenessRec.Membership.Decommissioned() &&
		newLiveness.Membership.Decommissioning() {
		// Marking a decommissioned node for decommissioning is a no-op.
		return false, nil
	}
```

Re-wrote the function a bit to encode this specific no-op transition.